feat(edit and update): edit and update tearsheet

jagathgj opened the pull request:
Contributes to #1260 

Wide Tearsheet, Narrow Tearsheet and FullPage are added on Edit and update section of Carbon for IBM Products PAL

#### What did you change?

Generated and added new stories for `EditTearsheet`, `EditTearsheetNarrow` and `EditFullPage`

#### How did you test and verify your work?

storybook

elycheea commented on packages/cloud-cognitive/src/components/EditFullPage/EditFullPage.mdx:
```diff
@@ -0,0 +1,227 @@
+
+## Overview
+
+When to use
+
+- In Carbon for IBM Products, the Full page create flow is the process the user
+  must complete in order to use a product or service.
+- In **IBM Cloud public**, it is the alternative to the Provisioning flow for a
+  non-billable service.
+
+When not to use
+
+- Don’t use the Full page option when the product or service is usable before
+  the current creation flow is initiated.
+- For **IBM Public Cloud**, do not use this when provisioning a service directly
+  from the Catalog, or when creating a non-billable resource that could happen
+  in a side panel/in context.
```
We should update this with [full page edit](https://pages.github.ibm.com/cdai-design/pal/patterns/edit/usage#full-page-edit)’s when to use section.

elycheea commented on packages/cloud-cognitive/src/components/EditFullPage/EditFullPage.stories.js:
```diff
@@ -0,0 +1,452 @@
+const defaultFullPageProps = {
+  nextButtonText: 'Next',
+  backButtonText: 'Back',
+  cancelButtonText: 'Cancel',
+  submitButtonText: 'Update',
```
```suggestion
  cancelButtonText: 'Cancel changes',
  submitButtonText: 'Save changes',
```

elycheea replied:
Actually, this could be okay as is. In the full page section of the PAL, we refer to an edit mode and settings page. If we keep Cancel/Update (or change it to Cancel/Save), this would probably represent an edit mode.

Maybe the stories for this would make more sense as “edit more” and “settings” to mirror the PAL guidance. The PAL site looks like it shows multiple sections of the same page rather than multiple steps like in `CreateFullPage`. Not sure if we’d need multiple steps to complete an edit? We can reach out to Shelley Dignum if we need more clarification.

elycheea commented on packages/cloud-cognitive/src/components/EditFullPage/EditFullPage.mdx:
```diff
@@ -0,0 +1,227 @@
+  getStoryId,
+  CodesandboxLink,
+} from '../../global/js/utils/story-helper';
+import { CreateFullPage } from '.';
```
Can we double check references to `CreateFullPage` are updated to `EditFullPage`?

elycheea commented on packages/cloud-cognitive/src/components/EditFullPage/EditFullPage.stories.js:
```diff
@@ -0,0 +1,452 @@
+  prepareStory,
+} from '../../global/js/utils/story-helper';
+import { action } from '@storybook/addon-actions';
+import { CreateFullPage } from '../CreateFullPage';
+import { CreateFullPageStep } from '../CreateFullPage/CreateFullPageStep';
```
We should make sure we’re using EditFullPage since we created a separate component for it.

jagathgj replied:
As per the instructions, we are not creating new components. We are creating only stories from existing components. Here these imports are required for creating story for EditFullPage.

elycheea commented on packages/cloud-cognitive/src/components/EditTearsheet/EditTearsheet.mdx:
```diff
@@ -0,0 +1,190 @@
+} from '../../global/js/utils/story-helper';
+import { CreateTearsheet } from '../CreateTearsheet';
+import { EditTearsheet } from '.';
+import createTearsheetGif from '../CreateTearsheet/storybook_assets/create_tearsheet.gif';
+import createTearsheetAnatomy from '../CreateTearsheet/storybook_assets/create_tearsheet_anatomy.jpg';
```
Should remove or update these screenshots.

elycheea commented on packages/cloud-cognitive/src/components/EditTearsheet/EditTearsheet.mdx:
```diff
@@ -0,0 +1,190 @@
+
+## Overview
+
+When to use
+
+- Use a wide tearsheet with creations that have two or more distinct steps.
+- Omit the progress indicator to use with single step creations that require
+  more space, such as those with interactive selections, diagrams, or tables.
+
+When not to use
+
+- Don’t use the wide tearsheet for low complexity create flows with only one
+  step.
```
Should also update this with the [EditTearsheet version](https://pages.github.ibm.com/cdai-design/pal/patterns/edit/usage#wide-tearsheet-edit) of this.

elycheea commented on packages/cloud-cognitive/src/components/EditTearsheet/EditTearsheet.stories.js:
```diff
@@ -0,0 +1,63 @@
+};
+
+const createTearsheetProps = {
+  title: 'Create topic',
```
```suggestion
  title: 'Edit topic',
```

elycheea reviewed (changes requested):
Looks like there’s several places where we’re referring to the Create flow version of the tearsheets or documentation. We should also make sure the titles refer to **Edit** instead of **Create**.

Since Edit and Create aren’t too different, I would consider simplifying the examples here a bit to match the PAL guidance a bit more (for example, in EditTearsheet (wide tearsheet edit) on the guidance, it lists that steps in the edit flow are less common so maybe we can also show examples that are simpler as well.

I recommend reaching out to the designer with any questions or for clarification though!

jagathgj commented on packages/cloud-cognitive/src/components/EditTearsheetNarrow/EditTearsheetNarrow.stories.js:
```diff
@@ -0,0 +1,293 @@
+} from '../../global/js/utils/story-helper';
+
+import { EditTearsheetNarrow } from '.';
+import { CreateTearsheetNarrow } from '../CreateTearsheetNarrow';
```
@elycheea Since we are referring from CreateTearsheetNarrow rather than creating a new component for EditTearsheetNarrow, we need this import to use as below

<img width="548" alt="image" src="https://user-images.githubusercontent.com/29351394/167810893-ab2b02a0-ca19-416b-99ba-708a58e119c2.png">

jagathgj commented:
@elycheea I have added suggested changes. Kindly take a look at it.

jagathgj commented:
@elycheea Fixed the buttons in EditTearsheet. Kindly review it and let me know if there needs any change.

elycheea commented on packages/cloud-cognitive/src/components/EditTearsheet/EditTearsheet.stories.js:
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright IBM Corp. 2022, 2022
+ * Copyright IBM Corp. 2021, 2021
```
These should be 2022.

elycheea reviewed (dismissed):
I think we still have a few instances of `CreateTearsheet` included. After speaking with @matthewgallo and @davidmenendez some more about it, we feel like we should keep the edit patterns pretty simple without the multistep, especially since steps in the edit flow aren’t very common.

Also make sure the years show 2022 since these are newly added files.

elycheea commented:
@davidmenendez will help re-review while I’m OOO! Thank you @jagathgj

jagathgj commented:
@elycheea @davidmenendez  As suggested, I deleted step hooks and their functions and simplified the design. Please have a look at it.

davidmenendez commented:
reviewing now. the tests appear to be failing though, so @jagathgj you should start by fixing those tests.

davidmenendez commented:
also it looks like there's an error in the build. please check the log https://app.netlify.com/sites/carbon-for-ibm-products/deploys/6284dc42ad2361000ac6da6b#L313

davidmenendez commented on packages/cloud-cognitive/src/components/EditTearsheet/index.js:
```diff
@@ -0,0 +1,10 @@
+
+export { EditTearsheet } from './EditTearsheet';
+export { EditTearsheetStep } from './EditTearsheetStep';
+export { EditTearsheetDivider } from './EditTearsheetDivider';
```
i think this is where the build error is coming from. i think you may have forgotten to add this file to the git push.

davidmenendez commented on packages/cloud-cognitive/src/components/EditTearsheet/EditTearsheet.js:
```diff
@@ -0,0 +1,345 @@
+    });
+
+    const sideNavItems = [
+      { label: 'Topic Name' },
```
there shouldn't be any hardcoded strings in the component. they should always be passed by the consumer.

davidmenendez commented on packages/cloud-cognitive/src/components/EditTearsheet/EditTearsheet.js:
```diff
@@ -0,0 +1,345 @@
+    ];
+
+    const influencer = (
+      <div className="tearsheet-stories__dummy-influencer-block">
```
i'm assuming this is storybook related? why is there a storybook specific class in the component?

davidmenendez commented on packages/cloud-cognitive/src/components/EditTearsheet/EditTearsheetStep.js:
```diff
@@ -0,0 +1,198 @@
+          // Pass through any other property values as HTML attributes.
+          ...rest
+        }
+        className={cx(blockClass, className, {
```
this is a personal preference thing, but i figured i'd point it out. i think it's good practice to avoid having logic or function calls in the HTML portion of react components. i would personally move this out to a variable prior to the return. something like `stepClassName` for instance to clean the HTML more clean.

davidmenendez commented on packages/cloud-cognitive/src/components/EditTearsheet/preview-components/MultiStepTearsheet.js:
```diff
@@ -0,0 +1,298 @@
+        open={open}
+        onHandleModalClick={handleModalClick}
+        onClose={clearCreateData}
+        onRequestSubmit={() =>
```
similar comment i made about having function calls like this in the HTML. i think you should consider moving these out into their own handlers. `onRequestSubmit={submitHandler}`

davidmenendez commented on packages/cloud-cognitive/src/components/EditTearsheet/preview-components/MultiStepTearsheet.js:
```diff
@@ -0,0 +1,298 @@
+/* eslint-disable react/prop-types */
```
what are `preview-components`?

davidmenendez commented on packages/cloud-cognitive/src/components/EditTearsheetNarrow/EditTearsheetNarrow.stories.js:
```diff
@@ -0,0 +1,293 @@
+      <Button onClick={() => setOpen(!open)}>
+        {open ? 'Close EditTearsheetNarrow' : 'Open EditTearsheetNarrow'}
+      </Button>
+      <CreateTearsheetNarrow
```
i think i'm confused about what's going on here. you're not actually rendering `<EditTearsheerNarrow />` anywhere?

davidmenendez reviewed (changes requested):
looks good overall. had a few comments and questions.

davidmenendez commented:
hey @jagathgj have you had a chance to review my comments?

jagathgj commented:
@davidmenendez I have committed the suggested changes. I'd appreciate it if you could take a look at it. When I try to test, I receive an error. I couldn't figure out what the problem is.

davidmenendez commented:
@jagathgj ok looking at the tests this is how i was able to get them to pass

* it's complaining about `window.matchMedia`. there are two options. either implement this workaround https://stackoverflow.com/questions/39830580/jest-test-fails-typeerror-window-matchmedia-is-not-a-function or do what Matt did in this PR https://github.com/carbon-design-system/ibm-cloud-cognitive/pull/2029
* it's also complaining about missing props for `<EditTearsheetForm>Form 1</EditTearsheetForm>`
* in `defaultProps` it's complaining about using the wrong prop name for `onClose` because you're using `onRequestClose`

fixing all those should get the tests running properly. i would also just reference the regular tearsheet tests to make sure you have everything set up properly.

jagathgj commented:
Thanks alot @davidmenendez. I will try these methods.

davidmenendez commented:
@jagathgj cool. please let me know when the changes have been made and you're ready for me to review again 👍

jagathgj commented:
@davidmenendez I have added suggested changes. Regarding the `window.matchMedia` error, I have added the workaround you have provided. Kindly review and let me know your feedback.

davidmenendez reviewed (approved):
thanks for the updates

jagathgj commented:
@davidmenendez Thanks David. I have fixed the lint issue. Kindly review and merge.

davidmenendez reviewed (approved):
TYVM
